// SPDX-FileCopyrightText: 2025 Carnegie Mellon University - LivingEdgeLab
//
// SPDX-License-Identifier: GPL-2.0-only

syntax = "proto3";
package steeleagle.protocol.services.mission_service;

import "steeleagle_sdk/protocol/common.proto";

/*
 * Used to start a new mission or stop an active mission
 */
service Mission {
  // Upload a mission for execution
  rpc Upload (UploadRequest)
	returns (steeleagle.protocol.common.Response) {}
  // Start an uploaded mission
  rpc Start (StartRequest)
	returns (steeleagle.protocol.common.Response) {}
  // Stop the current mission
  rpc Stop (StopRequest)
	returns (steeleagle.protocol.common.Response) {}
  // Send a notification to the current mission
  rpc Notify (NotifyRequest)
	returns (steeleagle.protocol.common.Response) {}
  // Set the mission telemetry stream parameters
  rpc ConfigureTelemetryStream (ConfigureTelemetryStreamRequest)
	returns (steeleagle.protocol.common.Response) {}
}

message MissionData {
  string content  = 1; // URI, either local or remote, of a mission file
  bytes map = 2; // kml object
}

message UploadRequest {
  steeleagle.protocol.common.Request request = 1;
  MissionData mission = 2; // Data of the target mission
}


message StartRequest {
  steeleagle.protocol.common.Request request = 1;
}


message StopRequest {
  steeleagle.protocol.common.Request request = 1;
}


message NotifyRequest {
  steeleagle.protocol.common.Request request = 1;
  int32 notify_code = 2; // Integer notification code, generated by the backend
}

message ConfigureTelemetryStreamRequest {
  steeleagle.protocol.common.Request request = 1;
  uint32 frequency = 2; // Target frequency for telemetry stream
}

message ConfigureTelemetryStreamResponse {
  steeleagle.protocol.common.Response response = 1;
}
